src/c4/Invert_Comm_Map.cc

```diff
@@ -17,74 +17,118 @@
+  return num_recv;
 }
 //---------------------------------------------------------------------------//
-// SCALAR version of invert_comm_map
+// SCALAR version of get_num_recv
 #elif defined(C4_SCALAR)
-void invert_comm_map(Invert_Comm_Map_t const &to_map,
-                     Invert_Comm_Map_t &from_map) {
-  Require(to_map.size() == 0u || (to_map.size() == 1u && to_map.at(0) > 0));
-  from_map.clear();
-  auto it = to_map.find(0);
-  if (it != to_map.end()) {
-    from_map[0] = it->second;
-  }
+int get_num_recv(Invert_Comm_Map_t::const_iterator first,
+                 Invert_Comm_Map_t::const_iterator last) {
+  return 0;
 }
 #else
 //---------------------------------------------------------------------------//
-// Default version of invert_comm_map, which throws an error.
-void invert_comm_map(Invert_Comm_Map_t const &, Invert_Comm_Map_t &) {
-  Insist(0, "invert_comm_map not implemented for this communication type!");
+// Default version of get_num_recv, which throws an error.
+int get_num_recv(Invert_Comm_Map_t::const_iterator first,
+                 Invert_Comm_Map_t::const_iterator last) {
+  Insist(0, "get_num_recv not implemented for this communication type!");
 }
 #endif // ifdef C4_MPI
 
+//---------------------------------------------------------------------------//
+void invert_comm_map(Invert_Comm_Map_t const &to_map,
+                     Invert_Comm_Map_t &from_map) {
+  const int my_proc = rtt_c4::node();
+  const int num_procs = rtt_c4::nodes();
+
+  // number of procs we will receive data from
+  const int num_recv = get_num_recv(to_map.begin(), to_map.end());
+
+  // request handle for the receives
+  std::vector<C4_Req> recvs(num_recv);
+
+  // the number of data elements to be received from each proc.  This data will
+  // ultimately be loaded into from_map, once we know the sending proc ids.
+  std::vector<size_t> sizes(num_recv);
+
+  // communication tag for sends/recvs
+  const int tag = 201;
+
+  // Posts the receives for the data sizes.  We don't yet know the proc numbers
+  // sending the data, so use any_source.
+  for (int i = 0; i < num_recv; ++i) {
+    receive_async(recvs[i], &sizes[i], 1, any_source, tag);
+  }
+
+  from_map.clear(); // empty whatever came in
+
+  // Send the data sizes and take care of on-proc map.
+  for (auto it = to_map.begin(); it != to_map.end(); ++it) {
+    Require(it->first >= 0);
+    Require(it->first < num_procs);
+    Require(it->second > 0);
+    if (it->first == my_proc) {
+      Check(from_map.find(my_proc) == from_map.end());
+      // on-proc map
+      from_map[my_proc] = it->second;
+    } else {
+      // we can ignore the request returned, because our send buffers are
+      // not shared, and we'll wait on the receives below.
+      send_async(&(it->second), 1, it->first, tag);
+    }
+  }
+
+  // Wait on the receives and populate the map
+  C4_Status status;
+  for (int i = 0; i < num_recv; ++i) {
+    recvs[i].wait(&status);
```

Comment:
C4 does provide a wait_all, which wraps MPi's wait_all.  But the wrapper ignores the status (it's passed MPI_STATUSES_IGNORE), which I need.  It'll need to be modified to support an array of C4_Status.  I don't need this for Invert_Comm_Map, so I'd rather not go down this route, until it's needed.  Same goes with wait_any().